_source/_posts/2022-07-21-jhipster-ci-cd.md

```diff
@@ -0,0 +1,865 @@
+    verbs: ['list']
+  - apiGroups: ['extensions', 'apps']
+    resources: ['daemonsets', 'deployments', 'deployments/scale', 'ingresses', 'replicasets', 'statefulsets']
+    verbs:
+      [
+        'create',
+        'delete',
+        'deletecollection',
+        'get',
+        'list',
+        'patch',
+        'update',
+        'watch',
+      ]
+  # These permissions are necessary for halyard to operate. We use this role also to deploy Spinnaker itself.
+  - apiGroups: ['']
+    resources: ['services/proxy', 'pods/portforward']
+    verbs:
+      [
+        'create',
+        'delete',
+        'deletecollection',
+        'get',
+        'list',
+        'patch',
+        'update',
+        'watch',
+      ]
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: jhipster-role-binding
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: jhipster-role
+subjects:
+  - namespace: demo
+    kind: ServiceAccount
+    name: jhipster-service-account
+---
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: jhipster-service-account
+  namespace: demo
+```
+
+Notice there are some differences with the Spinnaker service account. The k8s manifests generated by JHipster include the creation of `secrets`, `configmaps`, and `persistentvolumeclaims`, and the management of those Kubernetes resources must be granted to the service account used for deploying the JHipster k8s manifests.
+
+Create the Kubernetes `jhipster-service-account`:
+
+```shell
+CONTEXT=$(kubectl config current-context)
+
+kubectl apply --context $CONTEXT \
+    -f ./jhipster-service-account.yml
+
+
+TOKEN=$(kubectl get secret --context $CONTEXT \
+   $(kubectl get serviceaccount jhipster-service-account \
+       --context $CONTEXT \
+       -n demo \
+       -o jsonpath='{.secrets[0].name}') \
+   -n demo \
+   -o jsonpath='{.data.token}' | base64 --decode)
+
+kubectl config set-credentials ${CONTEXT}-token-user --token $TOKEN
+
+kubectl config set-context $CONTEXT --user ${CONTEXT}-token-user
+```
+
+Add the account to the Halyard configuration:
+
+```shell
+CONTEXT=$(kubectl config current-context)
+
+hal config provider kubernetes account add jhipster-gke-account \
+    --context $CONTEXT
+```
+
+### Add a docker registry account
+
+As the goal is to trigger the pipeline execution when a new image is pushed to DockerHub, you need to configure a docker-registry provider with Halyard.
+
+First, enable the docker-registry provider:
+
+```shell
+hal config provider docker-registry enable
+```
+
+The following `hal config` line will prompt for your password or access token if you have 2FA enabled. Make sure to replace `your-dockerhub-username`.
+
+```shell
+ADDRESS=index.docker.io
+USERNAME=your-dockerhub-username
+
+hal config provider docker-registry account add docker-account \
+    --address $ADDRESS \
+    --repositories $USERNAME/store $USERNAME/gateway \
+    --username $USERNAME \
+    --password
```

Comment:
I received an error at this point:

```
Your docker registry password:
+ Get current deployment
  Success
- Add the docker-account account
  Failure
Validation in halconfig:
- WARNING There is a newer version of Halyard available (1.45.0),
  please update when possible
? Run 'sudo update-halyard' to upgrade

Validation in default.provider.dockerRegistry.docker-account:
! ERROR Unable to fetch tags from the docker repository:
  mraible/store, Authentication failed: 401 Unauthorized
? Can the provided user access this repository?

! ERROR Unable to fetch tags from the docker repository:
  mraible/gateway, Authentication failed: 401 Unauthorized
? Can the provided user access this repository?

- WARNING None of your supplied repositories contain any tags.
  Spinnaker will not be able to deploy any docker images.
? Push some images to your registry.

! ERROR Unable to fetch tags from the docker repository:
  mraible/store, Authentication failed: 401 Unauthorized
? Can the provided user access this repository?

! ERROR Unable to fetch tags from the docker repository:
  mraible/gateway, Authentication failed: 401 Unauthorized
? Can the provided user access this repository?

- WARNING None of your supplied repositories contain any tags.
  Spinnaker will not be able to deploy any docker images.
? Push some images to your registry.

- Failed to add account docker-account for provider
  dockerRegistry.
```

Then I remembered that I should use an access token instead of my real password. When I did that, it all worked. You might want to mention this.

```
Your docker registry password:
+ Get current deployment
  Success
+ Add the docker-account account
  Success
Validation in halconfig:
- WARNING There is a newer version of Halyard available (1.45.0),
  please update when possible
? Run 'sudo update-halyard' to upgrade

+ Successfully added account docker-account for provider
  dockerRegistry.
Execution time: 6 s.
```

Reply:
I think it is mentioned. "prompt for your password or access token if you have 2FA enabled"